### PR TITLE
mpu4: Refactoring and functionality additions

### PR DESCRIPTION
**Pull request by james-wallace-ghub**
Added the ability to override the lamp current checks (will fix Lamp Drive errors)

Connect4 no longer requires a lamping hack

### CODE REVIEW

**Comment by james-wallace-ghub** [comment]
This is the reworked version of the previous PR, and I've added some other changes I was working on.

**Comment by james-wallace-ghub** [comment]
How do I refire the checks on this thing? I think the NMK machine problem stopped these, but I can't see how to restart them.

**Comment by startaq** [comment]
The checks will get executed on a new commit. As far as I can see there is no commit that solves the issue.

**Comment by DavidHaywood** [comment]
the failures seem unrelated to this PR (seems to be failing on NMK112 audio ?)

**Comment by startaq** [comment]
Oh, I didn't realize that the failure is unrelated. If you initially forked off from a broken compile you need to rebase your pull request (or just create a new pull request with your changes).

**Comment by james-wallace-ghub** [comment]
Tried fetching and merging at my end, see if that works

File: src/mame/barcrest/mpu4.cpp [hunk 1]
```diff
@@ -924,7 +929,15 @@ uint8_t mpu4_state::pia_ic5_portb_r()
...
+	if (m_dataport_loopback)
+	{
+		m_serial_data = state;
+		m_pia4->cb1_w(state);
+	}
+	else
+	{
+		m_dataport->write_txd(state);		
+	}
 }
```
Comment (cuavas) [inline src/mame/barcrest/mpu4.cpp:941]: Please get rid of this member entirely and just hook up the IC5 CA2 output to either the PIA4 CB1 input or the data logger directly in machine configuration.

File: src/mame/layout/connect4.lay [hunk 2]
```diff
@@ -3,47 +3,127 @@
...
-	<led7seg>
-			<color red="1.0" green="0.0" blue="0.0" />
-	</led7seg>
+
+	<element name="ledseg0">
+		<image statemask="0x01">
+		<data>
+			<![CDATA[
+				<svg xmlns="http://www.w3.org/2000/svg" width="40" height="60" viewBox="0 0 109.2832 165.8748">
+				<path fill="#ffffff" stroke="#ffffff" stroke-opacity="1" fill-opacity="1" stroke-width="0.265px" d="M 22.20542,9.1048101 28.73216,15.40339 90.121429,15.68542 98.067028,7.8827048 97.972444,6.3785695 91.634871,-0.01402362 H 30.245613 L 22.489191,7.8827048 Z" />
```
Comment (cuavas) [inline src/mame/layout/connect4.lay:12]: Do you actually need to stroke these paths?  They’re simple filled shapes – you’d be better off dropping the stroking and just adjusting the path so it follows the outlline of that you want filled.

File: src/mame/layout/connect4.lay [hunk 3]
```diff
@@ -640,77 +720,90 @@ license:CC0
...
+		<repeat count="16">
+			<param name="i" start="15" increment="-1"/>
+			<param name="x" start="0" increment="9"/>
+			<element name="vfd~i~" ref="vfd0">
+				<bounds x="~x~" y="120" width="9" height="14"/>
+			</element>
+		</repeat>
+
+	</view>
+</mamelayout>
```
Comment (cuavas) [inline src/mame/layout/connect4.lay:809]: You’re missing the final newline on this file.

**Review by cuavas — CHANGES_REQUESTED** [review]
This does look like a better approach than before.  Are you remembering to remove (or not add) the datalogger when appropriate?

**Comment by james-wallace-ghub** [comment]
> This does look like a better approach than before. Are you remembering to remove (or not add) the datalogger when appropriate?

At this point, this supplants rather than removes, as there's too many configurations to change. It does need a more complete replacement, but while everything is being touched I'm keeping it simple.

File: src/mame/layout/connect4.lay (hunk 3 above)
Comment (james-wallace-ghub) [reply src/mame/layout/connect4.lay:809]: Turns out my autorun of srcclean doesn't work any more, I'll resolve this with the rest

**Comment by DavidHaywood** [comment]
yeah, having / not having the datalogger will come later, that's quite a major refactor of machine structures in itself.

**Comment by james-wallace-ghub** [comment]
I can make the datalogger disappear from things that definitely don't use it, that's not a problem, see the current changes.

Ultimately, if you're defining m_dataport_loopback, things will work, and ultimately hardware doing that should call a no_bacta configuration.

However, a proper replacement for BACTA with a loopback device would have problems with the line handlers etc, and it's not a true loopback device, more a bodged cable by the look of it.

File: src/mame/barcrest/mpu4.cpp (hunk 1 above)
Comment (james-wallace-ghub) [reply src/mame/barcrest/mpu4.cpp]: Unfortunately, the fact that the edge detect is traced to an additional input port (IC4, port B, pin 7) means I had to keep serial detect around, I couldn't find an easy way of duplicating that signal.

**Comment by james-wallace-ghub** [comment]
@cuavas, I haven't marked as resolved because I haven't been able to comply with removing the ic5-ca2 member, but I'm happy to enter into more dialogue over this as it is a difficult thing to process in this case.
As far as I can tell, I cannot read a CB1 input pin as an output, which I would need to do to remove this member entirely.

**Review by cuavas — CHANGES_REQUESTED** [review]
You shouldn’t be adding `m_dataport_loopback` at all.  In machine configuration, just connect the output where it needs to go directly.

For data logger:
```
	m_pia5->ca2_handler().set(m_dataport, FUNC(bacta_datalogger_device::write_txd));
```

For loopback:
```
	m_pia5->ca2_handler().set(m_pia4, FUNC(pia6821_device::cb1_w));
```

You don’t need `m_serial_detect`, because the PIA exposes its output states.  You can access it directly and avoid redundant state:
```
	if (m_pia5->ca2_output())
		m_ic4_input_b |= 0x80;
```

**Comment by cuavas** [comment]
OK, this looks good to go for now.

**Pull request merged**